src/sentry/backup/dependencies.py

```diff
@@ -336,7 +381,55 @@ def dependencies() -> dict[NormalizedModelName, ModelRelations]:
                 # Sort the constituent sets alphabetically, so that we get consistent JSON output.
                 uniques=sorted(list(uniques), key=lambda u: ":".join(sorted(list(u)))),
             )
-    return model_dependencies_list
+
+    # Get a flat list of "root" models, then mark all of them as non-dangling.
+    relocation_root_models: list[NormalizedModelName] = []
+    for root_models in RelocationRootModels:
+        relocation_root_models.extend(root_models.value)
+    for model_name in relocation_root_models:
+        model_dependencies_dict[model_name].dangling = False
+
+    # Now that all `ModelRelations` have been added to the `model_dependencies_dict`, we can circle
+    # back and figure out which ones are actually dangling. We do this by marking all of the root
+    # models non-dangling, then traversing from every other model to a (possible) root model
+    # recursively. At this point there should be no circular reference chains, so if we encounter
+    # them, fail immediately.
+    def resolve_dangling(seen: Set[NormalizedModelName], model_name: NormalizedModelName) -> bool:
+        model_relations = model_dependencies_dict[model_name]
+        model_name = get_model_name(model_relations.model)
+        if model_name in seen:
+            raise RuntimeError(
+                f"Circular dependency: {model_name} cannot transitively reference itself"
+            )
+        if model_relations.dangling is not None:
+            return model_relations.dangling
+
+        # TODO(getsentry/team-ospo#190): Maybe make it so that `Global` models are never "dangling",
+        # since we want to export 100% of models in `ExportScope.Global` anyway?
+
+        seen.add(model_name)
+        for ff in model_relations.foreign_keys.values():
+            if not ff.nullable:
+                foreign_model_name = get_model_name(ff.model)
+                if not resolve_dangling(seen, foreign_model_name):
+                    # We only need one non-dangling reference to mark this model as non-dangling as
+                    # well.
+                    model_relations.dangling = False
+                    break
+
+        # If we successfully looped over all of the foreign keys without encountering a dangling
+        # reference, we know that this model is dangling as well.
+        if model_relations.dangling is None:
```

Comment:
nit: I think you can set this as true before we loop through all foreign keys

Reply:
Done.